wgsl/index.bs

```diff
@@ -4557,11 +4555,282 @@ TODO: *Stub*: Expression evaluation
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+<div class="note">Note: apart from two special vertices True and False, all vertices can be understood as having one of the following meanings:
+        - The control-flow at a specific program point must be uniform
+        - The value of a specific expression must be uniform
+        - A specific variable must always hold a uniform value
+
+        An edge can be understood as an implication from the statement corresponding to its source vertex to the statement corresponding to its target vertex.
+        In particular, to express that something must always be uniform (e.g. the control-flow at the call site of a derivative), we simply add an edge from True to the corresponding vertex.
+        Reciprocally, to express that we cannot ensure the uniformity of something (e.g. a variable which holds the thread id), we simply add an edge from the corresponding vertex to False
+</div>
+
+In more detail, here is the algorithm for analyzing a function:
+
+* Create some vertices called "True", "False", "CF_start", "CF_return" and if the function is non-void a vertex called "Value_return"
+* Create one vertex for each parameter of the function which we'll call "arg_i"
+* Walk over the AST of the function, adding vertices and edges to the graph following the rules of the next section.
+* Look at the *Behaviors* of the function body, using the rules for statements in the next section
+    * If it has a behavior which is neither Nothing nor Discard, then reject the program
+    * Else remember it for later.
+* Look at which vertices are reachable from "True"
+    * If this set includes the vertex "False", then reject the program.
+    * Remember for later whether this set includes "CF_start" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "CF_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* If "Value_return" exists, look at which vertices are reachable from it
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+
+Note: Anything which is not explicitly remembered, including the entire graph can be destroyed at this point. The little information that we do need to remember to analyze the caller is linear in the number of parameters of the function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
 
-### Divergence and reconvergence TODO ### {#divergence-reconvergence}
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
```

Comment:
This is the first mention of convergence or reconvergence. As I understand it, it is a synonym for uniformity and nonuniformity, but it's not explicitly said here. I happen to think that these terms are preferable due to the different meaning of uniform buffers, but regardless, they should either be defined or not used.

Reply:
oh this is going to be fun to specify! SPIR-V uses the notion of quad uniformity, while reconverting only at the subgroup scale.

Reply:
Point taken about terminology!  We can bikeshed when the text comes into better focus.   (We can use mangoes, bananas, grapes as needed.)